src/Common/MemoryWorker.cpp

```diff
@@ -0,0 +1,49 @@
+#include <Common/MemoryWorker.h>
+
+#include <Common/Jemalloc.h>
+#include <Common/MemoryTracker.h>
+#include <Common/formatReadable.h>
+
+namespace DB
+{
+
+#if USE_JEMALLOC
+MemoryWorker::MemoryWorker(uint64_t period_ms_)
+    : period_ms(period_ms_)
+{
+    background_thread = ThreadFromGlobalPool([this] { backgroundThread(); });
+}
+
+MemoryWorker::~MemoryWorker()
+{
+    {
+        std::unique_lock lock(mutex);
+        shutdown = true;
+    }
+    cv.notify_all();
+
+    if (background_thread.joinable())
+        background_thread.join();
+}
+
+void MemoryWorker::backgroundThread()
+{
+    JemallocMibCache<uint64_t> epoch_mib("epoch");
+    JemallocMibCache<size_t> resident_mib("stats.resident");
+    std::unique_lock lock(mutex);
+    while (true)
+    {
+        cv.wait_for(lock, period_ms, [this] { return shutdown; });
+        if (shutdown)
+            return;
+
+        epoch_mib.setValue(0);
+        Int64 resident = resident_mib.getValue();
+        MemoryTracker::setRSS(resident);
+        if (resident > total_memory_tracker.getHardLimit())
+            purgeJemallocArenas();
```

Comment:
I liked the soft limit we had.
Maybe we can update RSS right after purging arenas.

Initially, the idea to pugre arenas was implemented to allow run 2 queries with high memory one after one and do not get `Memory Limit` exception (because arenas are visible in RSS, from where we update MemoryTracker). I suspect that this case would not work now (but, probably, this is not very important, as long as it was not completely fixed anyway).

Reply:
I understand that part, but take into account that this thread is executing at a much higher rate than AsyncMetrics thread.

The old logic felt a bit racy because we purged memory while allowing new allocations so OOMs could happen easily.

IMO optimizing such things should be done after we make sure the server is stable and the global tracker does its main job of keeping the server alive.